tailassignment_oop.py

```diff
@@ -0,0 +1,376 @@
+        problem, with time parameter gamma
+
+        Parameters
+        ----------
+        gamma : float
+            Time/angle for applying hamiltonian.
+
+        """
+        
+        for r in range(self.R):
+            hr = 0.5 * self.CR[r] + 0.5 * self.mu * self.FR[:,r] @ (np.sum(self.FR,axis = 1) - 2)
+
+            self.qc.rz( gamma * hr, self.q_register[r])
+
+            for r_ in range(r+1,self.R):
+                Jrr_  = 0.5 * self.FR[:,r] @ self.FR[:,r_]
+
+                # Apply U(gamma), coupling part
+
+                self.qc.cx(self.q_register[r], self.q_register[r_])
+                self.qc.rz(gamma * Jrr_, self.q_register[r_])
+                self.qc.cx(self.q_register[r], self.q_register[r_])
+            
+    def createCircuit(self, params):
+
+        """
+        Implements the ciruit for the tail assignment problem
+        Parameters
+        ----------
+        params : array
+            variational parameters gamma_1 beta_1, ... , gamma_p beta_p
+
+        Returns
+        -------
+        qc : QuantumCircuit    
+
+        """
+
+        self.initial_state(self.R)
+        
+        gammas = params[::2]
+        betas  = params[1::2]
+
+        for d in range(self.depth):
+
+            gamma = gammas[d]
+            beta  = betas[d]
+
+            # Hamiltonian - cost + constraint
+            self.apply_hamiltonian(gamma)
+            # This is an equivalent implementation, but requires more gates.
+            # as the h-terms are not collected together
+            #self.apply_cost(gamma)
+            #self.apply_exco(gamma)
+
+            if self.options['usebarrier']:
+                self.qc.barrier()
+
+            # Apply mixer U(beta):
+            self.mix_states(beta)
+            
+            if self.options['usebarrier']:
+                self.qc.barrier()
+                
+        if "statevector" not in self.backend.name().split('_'):
+            # Do not measure at the end of the circuit if using a
+            # statevector simulation 
+            self.qc.measure(self.q_register,self.c_register)
+
+        return self.qc
+
+    def simulation_statistics(self, plot = True, savefig = None):
+        """
+        Do simulation again with optimal found parameters and 
+        return the success probability together with the average hamiltonian.
+
+        Parameters
+        ----------
+        plot : bool
+            If True, plot the results
+        savefig : string or None
+            If not None, save the figure with the filename/path given by savefig 
+
+        Returns
+        -------
+        SP : array
+            Success probability as a function of depth
+        C : array
+            Total cost as a function of depth.
+        """
+
+        SP = np.zeros(self.max_depth)
+        C  = np.zeros(self.max_depth)
+        
+        self.depth = 1
+        while self.depth <= self.max_depth:
+
+            qc  = self.createCircuit(self.params[f'xL_d{self.depth}'])
```

Comment:
does this assume state_vector_simulator?

Reply:
Yes. I can create an option for noisy simulation here.